### ISSUE TITLE
W3C CSS Validation

### ISSUE THREAD

**Issue by jbalsas** [liferay/clay#375]
A recent concern has been brought out to our attention that for Accessibility concerns and audits, a common requirement is that all used CSS should be valid. One available tool for this is the [W3C CSS Validation Service](http://jigsaw.w3.org/css-validator/).

Using it in the current `clay@2.0.0-ALPHA.11` [throws around 90 errors](http://jigsaw.w3.org/css-validator/validator?uri=http%3A%2F%2Fpat270.github.io%2Flexicon%2Fcss%2Fsite%2Fsite-atlas-font-awesome.css&profile=css3svg&usermedium=all&warning=1&vextwarning=true&lang=en) and more warnings.

It would be good if we could dig a little bit to see if we can:
- Manage a default 100% valid CSS3 build
- Create an alternative build that validates so it can be used in themes instead of the default one

@pat270, could you prioritize this a little bit to get a sense of how feasible this could be?

**Comment by jbalsas** [comment]
Moving to `3.x` label

**Comment by kresimir-coko** [comment]
What are we doing about this issue @pat270 @jbalsas ? Do we know of any actionable steps to take here? Have we done the steps listed above:
> Manage a default 100% valid CSS3 build
Create an alternative build that validates so it can be used in themes instead of the default one

**Comment by pat270** [comment]
@kresimir-coko There's not much we can do, most of the issues are related to vendor specific selectors and CSS variables. I suppose we can add a setting for CSS Variable code that lets you prevent it from being output.

**Comment by jbalsas** [comment]
Yeah, maybe we could do something to just validate our output (minus vendor) so at least we know that keeps us honest... @bryceosterhaus, @matuzalemsteles, thoughts on this?

**Comment by wincent** [comment]
I feel the same way about producing two builds in order to pass validation (what does it mean to have a "valid" output if the thing that's valid isn't the thing you ship?) as I do about having distinct builds for testing and running (the guarantees the test suite offers about the behavior of the production code are diluted when you don't actually test the same code as you deploy).

![grumpy-old-man](https://user-images.githubusercontent.com/7074/92014515-a5517600-ed4f-11ea-8938-e8c95ce60d8f.gif)

**Comment by jbalsas** [comment]
> I feel the same way about producing two builds in order to pass validation

I feel you grumpy!!

> [...] what does it mean to have a "valid" output if the thing that's valid isn't the thing you ship? [...]

I proposed this having in mind that we're on a (long) path to remove Bootstrap 🤞. With that in mind, I was thinking that the more we removed, the closer our own thing gets to what we ship... so keeping it validated now would mean we wouldn't have to work on it after the fact.

That being said, I'm fine putting this off until we can actually ship a valid artifact, so no hard requirement on my side.

**Comment by matuzalemsteles** [comment]
I didn't quite understand, I'm confused haha ​​😆! But if this is a problem with the dependencies we have and not necessarily with Clay, maybe evaluating just Clay would not be a start? If this is a problem with Bootstrap maybe it is worth starting to work on it in 4.x? @pat270 worked removing things from Bootstrap in the codebase, but do we still have to refine them to be valid?

**Comment by bryceosterhaus** [comment]
If we think it's helpful, I have no issues adding some sort of script to validate our custom css. Personally I don't have a huge preference. I did see [this package](https://www.npmjs.com/package/css-validator) that we might be able to leverage but not sure if it would cover the case @pat270 brings up though.

**Comment by matuzalemsteles** [comment]
> The problem is the CSS spec. The W3C validation site doesn't include https://www.w3.org/TR/css-variables-1/. There are also lots of vendor specific selectors and values we use to style our components that are not in the spec (e.g., ::-webkit-slider-thumb, ::-moz-range-thumb, and ::-ms-thumb).

Hmm, now I'm wondering if this is an updated specification 🤷‍♂️
